File: kotlin/lib/src/main/kotlin/desertroad/solar/Moment.kt
package desertroad.solar

import desertroad.solar.internal.Angle.Companion.degrees

/**
 * Contains information about the position of the Sun at a specific [time] from the observer's location.
 *
 * All angular measurements are in degrees.
 */
abstract class Moment internal constructor(@JvmField val localSolar: LocalSolar, @JvmField val time: Long) {
    abstract val altitude: Double

    abstract val azimuth: Double

    val zenith get() = 90.0 - altitude

    val phase by lazy { Phase.fromAltitude(altitude.degrees) }

    override fun toString() =
        "Moment(time=[%tc], phase=%s, altitude=%.2f, azimuth=%.2f)".format(time, phase, altitude, azimuth)
}

File: dart/lib/src/moment.dart
import 'package:local_solar/src/angle.dart';
import 'package:local_solar/src/local_solar.dart';
import 'package:local_solar/src/phase.dart';

class Moment {
  final LocalSolar localSolar;
  final DateTime time;
  final double Function() _altitude;
  final double Function() _azimuth;

  Moment(this.localSolar, this.time, this._altitude, this._azimuth);

  double get azimuth => _azimuth();
  double get altitude => _altitude();

  double get zenith => 90.0 - altitude;

  late final phase = Phase.fromAltitude(altitude.degrees);

  @override
  String toString() {
    return 'Moment{time: $time, altitude: $altitude, azimuth: $azimuth, phase: $phase}';
  }
}
